clippy_lints/src/clones_into_boxed_slices.rs

```diff
@@ -0,0 +1,204 @@
+    /// ```no_run
+    /// let boxed: Box<str> = Box::from("example");
+    /// ```
+    #[clippy::version = "1.93.0"]
+    pub CLONES_INTO_BOXED_SLICES,
+    perf,
+    "Cloning then converting into boxed slice instead of using Box::from"
+}
+declare_lint_pass!(ClonesIntoBoxedSlices => [CLONES_INTO_BOXED_SLICES]);
+
+fn count_refs(mut expr_ty: Ty<'_>) -> i64 {
+    let mut count = 0;
+    while let ty::Ref(_, inner, _) = expr_ty.kind() {
+        expr_ty = *inner;
+        count += 1;
+    }
+    count
+}
+
+// Shows the lint with a suggestion using the given parts
+// Assures that the inner argument is correctly ref'd/deref'd in the suggestion based on needs_ref
+fn show_lint(
+    cx: &LateContext<'_>,
+    full_span: Span,
+    mut inner: &Expr<'_>,
+    needs_ref: bool,
+    sugg_prefix: Option<&str>,
+    placeholder: &str,
+    sugg_suffix: Option<&str>,
+) {
+    let mut applicability = Applicability::MachineApplicable;
+
+    while let ExprKind::AddrOf(BorrowKind::Ref, _, expr) | ExprKind::Unary(UnOp::Deref, expr) = inner.kind {
+        inner = expr;
+    }
+
+    let mut sugg = Sugg::hir_with_context(cx, inner, full_span.ctxt(), placeholder, &mut applicability);
+
+    let inner_ty = cx.typeck_results().expr_ty(inner);
+    let mut ref_count = count_refs(inner_ty);
+    if needs_ref {
+        if ty_is_slice_like(cx, inner_ty.peel_refs()) {
+            ref_count -= 1;
+        } else {
+            // Inner argument is in some kind of Rc-like object, so it should be addr_deref'd to get a reference
+            // to the underlying slice
+            sugg = sugg.addr_deref();
+        }
+    }
+    while ref_count > 0 {
+        sugg = sugg.deref();
+        ref_count -= 1;
+    }
+    while ref_count < 0 {
+        sugg = sugg.addr();
+        ref_count += 1;
+    }
+
+    span_lint_and_sugg(
+        cx,
+        CLONES_INTO_BOXED_SLICES,
+        full_span,
+        "clone into boxed slice",
+        "use",
+        format!(
+            "Box::from({}{}{})",
+            sugg_prefix.unwrap_or_default(),
+            sugg,
+            sugg_suffix.unwrap_or_default()
+        ),
+        applicability,
+    );
+}
+
+// Is the given type a slice, path, or one of the str types
+fn ty_is_slice_like(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    ty.is_slice()
+        || ty.is_str()
+        || ty.is_diag_item(cx, sym::cstr_type)
+        || ty.is_diag_item(cx, sym::Path)
+        || ty.is_diag_item(cx, sym::OsStr)
+}
+
+// Checks if an expression is one of the into_boxed_... methods preceded by a clone-like function
+// Then shows the lint with a suggestion that depends on the types of the inner argument and the
+// resulting Box
+impl<'tcx> LateLintPass<'tcx> for ClonesIntoBoxedSlices {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, second_method: &'tcx Expr<'_>) {
+        // Is the second method into_boxed_...?
+        if let ExprKind::MethodCall(second_method_path, first_method, _, _) = second_method.kind
```

Comment:
This would certainly be simplified by putting this lint into `methods`. The matching is basically already done there.